torchtitan/models/llama_vision/vision_encoder.py

```diff
@@ -0,0 +1,552 @@
+    in parallelism. The output of this module is slightly different from PyTorch's
+    Conv2D. While the latter has output to be (*, out_channels, h_out, w_out),
+    this module just has output to be (*, h_out * w_out, out_channels).
+
+    More details can be found in the PyTorch documentation:
+    https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html
+
+    We might later column-wise parallelize over unfolded. And we assume that
+    the shape of the kernel is always square.
+
+    Arguments:
+        in_channels: Input channels.
+        out_channels: Output channels.
+        kernel_size: Size of convolution kernel.
+        stride (default 1): Stride for convolution.
+        bias (default False): Use bias in Conv2d.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: int,
+        stride: int,
+        bias: Optional[bool] = False,
+    ) -> None:
+        super().__init__()
+        self._unfold = torch.nn.Unfold(
+            kernel_size=(kernel_size, kernel_size), stride=stride
+        )
+        self._linear = torch.nn.Linear(
+            in_channels * kernel_size * kernel_size,
+            out_channels,
+            bias=bias,
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Input: (bsz, in_channels, width, height)
+        # Output: (bsz, in_channels * kernel_size * kernel_size, num_tokens)
+        x = self._unfold(x)
+        x = x.permute(0, 2, 1)
+        # Output: (bsz, num_tokens, out_channels), when stride = kernel_size,
+        # num_tokens = grid ** 2 and out_channels is emd_dim.
+        return self._linear(x)
+
+
+class VitTransformerBlock(nn.Module):
+    def __init__(
+        self,
+        model_args: ModelArgs,
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
+    ):
+        super().__init__()
+        self.attn = Attention(model_args)
+        self.ln_attn = Fp32LayerNorm(model_args.encoder_embed_dim, eps=1e-5)
+        self.mlp = FeedForward(
+            dim=model_args.encoder_embed_dim,
+            hidden_dim=4 * model_args.encoder_embed_dim,
+            multiple_of=model_args.multiple_of,
+            ffn_dim_multiplier=model_args.ffn_dim_multiplier,
+            activation=model_args.activation,
+            enable_w3=False,
+        )
+        self.ln_mlp = Fp32LayerNorm(model_args.encoder_embed_dim, eps=1e-5)
+        self.attn_scale = attn_scale or nn.Identity()
+        self.mlp_scale = mlp_scale or nn.Identity()
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+    ):
+        bsz, seq_len, emd_dim = x.shape
+        x = x.view(bsz * seq_len, emd_dim)
+        x = x + self.attn_scale(self.attn(x=self.ln_attn(x), freqs_cis=None))
+        x = x + self.mlp_scale(self.mlp(self.ln_mlp(x)))
+        return x.view(bsz, seq_len, emd_dim)
+
+
+class LearnableTransformerBlock(nn.Module):
```

Comment:
Can this be coalesced with `VitTransformerBlock`? It seems the only difference is `attn_scale` and `mlp_scale`?

Reply:
yes, sorry I forgot to delete this one.